finality-verifiers/grandpa/src/lib.rs

```diff
@@ -0,0 +1,2046 @@
+        // 2. As this is not PoW, we must ensure there is a valid GrandpaJustification for the last header of what we're submitting. This can be seen as ensuring the correct fork is selected
+        // 3. The justification verifies a header that follows the linkage rule of the range
+
+        // For efficiency we check the the justification first. If it's invalid, we can skip the rest
+        let (signed_hash, signed_number) = (signed_header.hash(), signed_header.number());
+        let authority_set =
+            <CurrentAuthoritySet<T, I>>::get().ok_or_else(|| Error::<T, I>::InvalidAuthoritySet)?;
+
+        let set_id = authority_set.set_id;
+        // °°°°° Begin Check: #2 °°°°°
+        verify_justification_single::<T, I>(
+            &justification,
+            signed_hash,
+            *signed_number,
+            authority_set,
+            gateway_id,
+        )?;
+        // °°°°° Checked: #2 °°°°°°
+
+        // check for authority set update and enact if available.
+        let _enacted =
+            try_enact_authority_change_single::<T, I>(&signed_header, set_id, gateway_id)?;
+
+        // We get the latest buffer_index, which maps to the next header we can overwrite, and the index where we insert the verified header
+        let mut buffer_index =
+            <MultiImportedHashesPointer<T, I>>::get(gateway_id).unwrap_or_default();
+
+        // °°°°° Begin Check: #1 °°°°°
+        for header in range {
+            if best_finalized.hash() == *header.parent_hash() {
+                // write header to storage if correct
+                write_and_clean_header_data::<T, I>(
+                    gateway_id,
+                    &mut buffer_index,
+                    &header,
+                    header.hash(),
+                    false,
+                )?;
+
+                best_finalized = header;
+            } else {
+                // if anything fails here, noop!
+                return Err(Error::<T, I>::InvalidRangeLinkage.into())
+            }
+        }
+        // °°°°° Check Success: #1 °°°°°
+
+        // °°°°° Begin Check: #3 °°°°°
+        if best_finalized.hash() == *signed_header.parent_hash() {
+            // write header to storage if correct
+            write_and_clean_header_data::<T, I>(
+                gateway_id,
+                &mut buffer_index,
+                &signed_header,
+                signed_hash,
+                true,
+            )?;
+        } else {
+            return Err(Error::<T, I>::InvalidJustificationLinkage.into())
+        }
+        // °°°°° Check Success: #3 °°°°°
+        // Proof success! Submitted header range valid
+
+        // Update pointer
+        <MultiImportedHashesPointer<T, I>>::insert(gateway_id, buffer_index);
+
+        // ToDo use unique_saturated_into() here.
+        let height: usize = signed_number.as_();
+        let block_height: u64 = height.try_into().unwrap();
```

Comment:
Todo & panic, same below

Reply:
I changed this to handle an error, but am unable to use `unique_saturated_into()`.